Libraries/LibWasm/AbstractMachine/BytecodeInterpreter.cpp

```diff
@@ -391,1286 +2515,359 @@ double BytecodeInterpreter::read_value<double>(ReadonlyBytes data)
+            group_end = max(group_end, interval->end);
+        }
+
+        expire_old_intervals(group_start);
+
+        Dispatch::RegisterOrStack reg = Dispatch::RegisterOrStack::Stack;
+
+        bool group_forced_to_stack = false;
+        for (auto* interval : group) {
+            if (interval->forced_to_stack) {
+                group_forced_to_stack = true;
+                break;
+            }
+        }
+
+        if (!group_forced_to_stack) {
+            Array<bool, Dispatch::CountRegisters> used_regs;
+            used_regs.fill(false);
+
+            for (auto const& active_entry : active_by_end) {
+                if (active_entry.reg != Dispatch::RegisterOrStack::Stack)
+                    used_regs[to_underlying(active_entry.reg)] = true;
+            }
+
+            for (u8 r = 0; r < Dispatch::CountRegisters; ++r) {
+                if (used_regs[r])
+                    continue;
+
+                auto can_assign = true;
+                for (auto* interval : group) {
+                    auto interval_root = final_roots[interval->value_id];
+
+                    for (auto const* other_interval : reg_intervals[r]) {
+                        if (interval_root == final_roots[other_interval->value_id])
+                            continue;
+                        if (interval->end >= other_interval->start && other_interval->end >= interval->start) {
+                            can_assign = false;
+                            break;
+                        }
+                    }
+                    if (!can_assign)
+                        break;
+                }
+
+                if (can_assign) {
+                    reg = static_cast<Dispatch::RegisterOrStack>(r);
+                    active_by_end.insert(group_end, { entry.key, group_end, reg });
+
+                    for (auto* interval : group)
+                        reg_intervals[r].append(interval);
+                    break;
+                }
+            }
         }
+
+        for (auto* interval : group)
+            value_alloc.set(interval->value_id, reg);
     }
 
-    BytecodeInterpreter::interpret_instruction(configuration, ip, instruction);
+    for (size_t i = 0; i < result.dispatches.size(); ++i) {
+        auto& dispatch = result.dispatches[i];
+        auto input_ids = instr_to_input_values.get(i).value_or({});
 
-    if (post_interpret_hook) {
-        auto result = post_interpret_hook(configuration, ip, instruction, *this);
-        if (!result) {
-            m_trap = Trap::from_string("Trapped by user request");
-            return;
+        for (size_t j = 0; j < input_ids.size(); ++j) {
+            auto reg = value_alloc.get(input_ids[j]).value_or(Dispatch::RegisterOrStack::Stack);
+            dispatch.sources[j] = reg;
         }
+
+        auto output_id = instr_to_output_value.get(i).value_or(NumericLimits<size_t>::max());
+        if (output_id != NumericLimits<size_t>::max())
```

Comment:
is `output_id` supposed to naturally - read w/o the above fallback - reach this point being size_t::max?

Reply:
hmmm, I don't think so, will switch to has_value.